rclcpp_lifecycle/test/test_lifecycle_timer.cpp

```diff
@@ -0,0 +1,165 @@
+    wall_timer_called_ = false;
+    timer_ = node_->create_lifecycle_timer(
+      10ms, std::bind(&TestLifecycleTimer::timer_callback, this));
+    wall_timer_ = node_->create_lifecycle_wall_timer(
+      10ms, std::bind(&TestLifecycleTimer::wall_timer_callback, this));
+
+    spinner_ = std::thread(&TestLifecycleTimer::spin, this);
+  }
+
+  void TearDown()
+  {
+    {
+      std::lock_guard<std::mutex> guard(shutdown_mutex_);
+      rclcpp::shutdown();
+    }
+    spinner_.join();
+  }
+
+protected:
+  void spin()
+  {
+    while (true) {
+      {
+        std::lock_guard<std::mutex> guard(shutdown_mutex_);
+        if (!rclcpp::ok()) {
+          break;
+        }
+        rclcpp::spin_some(node_->get_node_base_interface());
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+    }
+  }
+
+  void timer_callback()
+  {
+    timer_called_ = true;
+  }
+
+  void wall_timer_callback()
+  {
+    wall_timer_called_ = true;
+  }
+
+  using CallbackT = std::shared_ptr<rclcpp_lifecycle::LifecycleTimer
+      <std::_Bind<void(TestLifecycleTimer::* (TestLifecycleTimer *))()>,
+      (void *) nullptr>>; // NOLINT
+  std::shared_ptr<rclcpp_lifecycle::LifecycleNode> node_;
+  CallbackT timer_;
+  CallbackT wall_timer_;
+  std::atomic<bool> timer_called_;
+  std::atomic<bool> wall_timer_called_;
+
+  std::mutex shutdown_mutex_;
+  std::thread spinner_;
+};
+
+TEST_F(TestLifecycleTimer, timer_managed_by_node) {
+  // Test that the timers transition with the node.
+  auto success = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+  auto reset_key = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  auto ret = reset_key;
+
+  EXPECT_EQ(State::PRIMARY_STATE_UNCONFIGURED, node_->get_current_state().id());
+  node_->trigger_transition(
+    rclcpp_lifecycle::Transition(Transition::TRANSITION_CONFIGURE), ret);
+  ASSERT_EQ(success, ret);
+  ret = reset_key;
+  node_->trigger_transition(
+    rclcpp_lifecycle::Transition(Transition::TRANSITION_ACTIVATE), ret);
+  ASSERT_EQ(success, ret);
+  ret = reset_key;
+  EXPECT_TRUE(timer_->is_activated());
+  EXPECT_TRUE(wall_timer_->is_activated());
+
+  node_->trigger_transition(
+    rclcpp_lifecycle::Transition(Transition::TRANSITION_DEACTIVATE), ret);
+  ASSERT_EQ(success, ret);
+  ret = reset_key;
+  (void)ret;  // Make clang happy
+  EXPECT_FALSE(timer_->is_activated());
+  EXPECT_FALSE(wall_timer_->is_activated());
+}
+
+TEST_F(TestLifecycleTimer, timer_callback_only_when_active) {
+  // Test that the callback only gets called when the timer is active.
+  auto success = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
+  auto reset_key = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;
+  auto ret = reset_key;
+
+  // Timer should not have been called yet
+  EXPECT_EQ(timer_called_, false);
+  EXPECT_EQ(wall_timer_called_, false);
```

Comment:
i think there probably are failing since the timer will autostart w/o state checking.